### ISSUE TITLE
multi_open error

### ISSUE THREAD

**Issue by wkliao** [DataLib-ECP/vol-log-based#38]
```
*** TESTING CXX    multi_open: Creating files ----
In H5VL_log_file_open at H5VL_log_file.cpp:303: File already exist
```
See github action log in:
https://github.com/DataLib-ECP/vol-log-based/actions/runs/3434873711/jobs/5726602956#step:7:1643

**Comment by yzanhua** [comment]
According to [this line](https://github.com/DataLib-ECP/vol-log-based/blob/8b68e0f460edb3b2955c6dfcc7ef5c36d02a9301/tests/testcases/multi_open.cpp#L54), this multi_open test program is expected to fail here.

We can silence the error trace from HDF5 by adding `H5E_BEGIN_TRY` and `H5E_END_TRY` to the multi_open test program:
```c++
H5E_BEGIN_TRY{
        fid2 = H5Fopen (file_name, H5F_ACC_RDONLY, faplid);
}H5E_END_TRY;
```

However, using `H5E_BEGIN_TRY` and `H5E_END_TRY`, we will still see error messages from Log VOL. For example, [this line](https://github.com/DataLib-ECP/vol-log-based/actions/runs/3434873711/jobs/5726602956#step:7:1648) is the error message from Log VOL. To silence this error message, we need to:
1. modify Log VOL so that no error messages are printed inside the H5E_BEGIN_TRY/H5E_END_TRY block. Or
2. modify Log VOL so that error messages are only printed in DEBUG mode.

I don't think we should do option 2 above because the (I think) error messages should still be printed even under release/production mode, and  HDF5 also prints error messages during production mode.

Option 2 above is kind of related to the issue of Log VOL's warning messages when running `ncdump`. The issue occurs because `ncdump` tries to open a non-existent attribute and Log VOL treats this as an error. So it prints out the "error" message. However, using `ncdump` with native HDF5 will not produce any warning/error messages. 

I need to further investigate how `ncdump` + native HDF5 achieves that.

**Comment by wkliao** [comment]
One important question is "does Log VOL support opening the same file multiple times?"

Please do not mix the use of the terms errors and warnings. It is very common for
users to query whether an object exists (e.g. attribute or dataset). This can be
simply done by calling H5Aopen and then checking the returned code.
The failure message of H5Aopen should be considered as a warning and
only printed in the debug mode.

In PnetCDF, we do not print any warning messages, except for fatal errors.
It is users' responsibility to check all returned codes.

**Comment by yzanhua** [comment]
Currently, Log VOL does not support opening the same file multiple times. 
We made this decision partially because the underlying native VOL  has undefined behavior if a file is opened multiple times (see the "Attention" section of [H5Fopen document](https://docs.hdfgroup.org/hdf5/develop/group___h5_f.html#title33))

I'll make sure these warning messages only be printed in the debug mode.

**Comment by wkliao** [comment]
If we decide not to support multiple open of the same file, then can Log VOL detects the 2nd call and return an error?

**Comment by yzanhua** [comment]
Yes.  And this multi_open test case is to test this.

**Comment by wkliao** [comment]
The current error message is
```
H5VL_log_file.cpp:303: File already exist
```
It does not properly reflect the limitation we impose.
Consider "The same file has been opened. Log VOL currently does not support multiple opens."

**Comment by yzanhua** [comment]
I'll change the message to an informative one as you suggested.

**Comment by wkliao** [comment]
One question.
When stacking the Log VOL on top of Cache VOL, will this multiple file open happen?

**Comment by yzanhua** [comment]
When stacking the Log VOL on top of Cache VOL, internally Log VOL needs to return multiple file objects (`H5VL_log_file_t`) to the HDF5 library because of the possible [H5VL_log_wrap_object](https://github.com/DataLib-ECP/vol-log-based/blob/f9b18d00643de60a4773ac157d797cead10740cb/src/H5VL_log_wrap.cpp#L106) calls.

This is similar to multiple file open in the sense that multiple file objects (`H5VL_log_file_t`) need to be returned but we don't actually need to open the same file multiple times. Right now we return the same (pointer of) the file object but added a [new field](https://github.com/DataLib-ECP/vol-log-based/blob/f9b18d00643de60a4773ac157d797cead10740cb/src/H5VL_log_file.hpp#L76) to keep track of the information needed by H5VL_log_wrap_object.

Currently, the implementation is tested and works fine for the following:
1. Log VOL on top of native VOL (`make check` tests this)
2. Log VOL on top of Cache VOL (tested personally, not tests in `make check`)
3. Log VOL on top of Cache VOL, on top of Async VOL

**Comment by wkliao** [comment]
If this issue has been resolved in a PR, then refer the PR number and close this issue.

**Comment by yzanhua** [comment]
Resolved in PR #39.
